### PR TITLE
Refactor state store bulk operations

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
# Description

This is a companion PR for dapr/components-contrib#2747 that refactors how bulk state store operations are performed

1. Changes the default BulkSet/BulkDelete implementation to use transactions (i.e. the Multi method) when the component supports that. This makes sure that operations are performed atomically when possible, and should allow for better performance. When the component is not transactional, it falls back to performing them in sequence.
2. Added a BulkGet implementation in the default bulk store.
   - Previously, the runtime implemented this in [three](https://github.com/dapr/dapr/blob/fa7a8d2989b701ceaa9bfba656fc98dd5c3cce1f/pkg/actors/actors.go#L1522) [different](https://github.com/dapr/dapr/blob/fa7a8d2989b701ceaa9bfba656fc98dd5c3cce1f/pkg/grpc/api.go#L611) [places](https://github.com/dapr/dapr/blob/fa7a8d2989b701ceaa9bfba656fc98dd5c3cce1f/pkg/http/api.go#L682), which had some differences and were hard to maintain (see bugs like [these](https://github.com/dapr/dapr/pull/6134)). At least one of them also had a TODO saying it should be refactored into components-contrib
3. The shared implementation gets all elements in sequence, but components that can perform bulk get operations are encouraged to do so.
   - Some components have the implementation added in this PR, others were added [previously](https://github.com/dapr/components-contrib/pull/2704) (SQLite, Postgres, CockroachDB)
4. Another change is in the internal interface used by Dapr to send the list of transactional requests, so we don't need to perform costly conversions using `mapstructure` anymore.

From a runtime perspective, there's one change in behavior when multiple keys being fetched individually.

- Previously, resiliency was applied to each individual Get operation, so if a single key failed, it would have been retried. Now, Resiliency is applied to the entire bulk get operation. This only impacts cases where there's a partial error.
- This is arguably more correct, for 2 reasons:
  1. First, components should be encouraged to implement native BulkGet, where multiple keys are retrieved in a single operation. This has already started in components-contrib, with 5 components updated so far.
  2. Second, when operations are retried individually, it's more likely that data will not be consistent if it's being written by someone else at the same time.

## Issue reference

Related to #6121

### CODE REVIEW

File: pkg/runtime/wfengine/wfengine_test.go [hunk 1]
```diff
@@ -19,183 +19,31 @@ package wfengine_test
...
 	"github.com/dapr/dapr/pkg/resiliency"
 	"github.com/dapr/dapr/pkg/runtime/wfengine"
+	daprt "github.com/dapr/dapr/pkg/testing"
 	"github.com/dapr/kit/logger"
 )
 
 const testAppID = "wf-app"
 
-// The fake state store code was copied from actors_test.go.
-// TODO: Find a way to share the code instead of copying it, if it makes sense to do so.
```
Comment (ItalyPaleAle) [inline pkg/runtime/wfengine/wfengine_test.go:47]: Reviewers: this code was moved to the testing package to address this TODO and reduce code duplication

**Comment by cgillum** [comment]
Is there anything else remaining in this PR that would prevent it from being reviewed/merged?

**Comment by ItalyPaleAle** [comment]
No, @cgillum . This PR is ready and it is necessary to be able to pin the latest components-contrib

CC: @dapr/maintainers-dapr

**Review by mukundansundar — COMMENTED** [review]
> Changes the default BulkSet/BulkDelete implementation to use transactions (i.e. the Multi method) when the component supports that. This makes sure that operations are performed atomically when possible, and should allow for better performance. When the component is not transactional, it falls back to performing them in sequence.

@ItalyPaleAle The main question that I have here is, is it expected that Bulk operations are going to be transactional?

> Previously, resiliency was applied to each individual Get operation, so if a single key failed, it would have been retried. Now, Resiliency is applied to the entire bulk get operation. This only impacts cases where there's a partial error.

Isn't this the case for BulkSet/BulkDelete according to the previous quote that they are going to use transactions .. In which case resiliency is going to change in behavior for set, get and delete to be based on transactions and not individual events.

**Comment by ItalyPaleAle** [comment]
@mukundansundar it's a good question.

This doesn't change the behavior: to end users, bulk operations are still performed "in bulk", and whether they are transactional or not is an internal implementation detail that we will not exposed.

A bunch of components have always implemented bulk set / bulk delete operations using transactions. Here are some examples from 1.10:

- MySQL: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/mysql/mysql.go#L578-L602
- Postgres: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/postgresql/postgresdbaccess.go#L231-L255
- SQLite: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlite/sqlite.go#L105-L114
- SQL Server: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlserver/sqlserver.go#L615-L634
- Oracle: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/oracledatabase/oracledatabase.go#L90-L93

What I did was moving the code that was repeated in components and standardizing it for all state stores that are transactional.

About Resiliency:

- In the case of BulkSet and BulkDelete, nothing changes actually, because before Resiliency was applied to the "BulkSet" invocation and not each individual Set. The loop happened inside the DefaultBulkStore that was unaware of resiliency. This would actually create some interesting issues when not using transactions and doing a BulkSet with Etags, if some items are set and some are not
- In the case of BulkGet there is a behavior change with Resiliency as explained in the first message.

File: dapr/proto/components/v1/state.proto [hunk 2]
```diff
@@ -264,7 +269,6 @@ message BulkStateItem {
 
 message BulkGetResponse {
   repeated BulkStateItem items = 1;
-  bool got = 2;
```
[the remark this answers is not in the document]
Comment (philliphoff) [reply dapr/proto/components/v1/state.proto:267]: I wasn't a fan of that property anyway. :-) I suppose that means the existing SDKs won't be usable (for bulk state store operations) until they're updated.  I'll add issues in those repos to do that.

[the remark this answers is not in the document]
Comment (philliphoff) [reply dapr/proto/components/v1/state.proto:267]: @ItalyPaleAle This change is targeting 1.11 or 1.12?

[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply dapr/proto/components/v1/state.proto:267]: We need this in 1.11. This PR is actually blocking components-contrib to be updated in dapr/dapr

**Review by daixiang0 — COMMENTED** [review]
/ok-to-test

**Comment by artursouza** [comment]
> @mukundansundar it's a good question.
> 
> This doesn't change the behavior: to end users, bulk operations are still performed "in bulk", and whether they are transactional or not is an internal implementation detail that we will not exposed.
> 
> A bunch of components have always implemented bulk set / bulk delete operations using transactions. Here are some examples from 1.10:
> 
> * MySQL: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/mysql/mysql.go#L578-L602
> * Postgres: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/postgresql/postgresdbaccess.go#L231-L255
> * SQLite: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlite/sqlite.go#L105-L114
> * SQL Server: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlserver/sqlserver.go#L615-L634
> * Oracle: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/oracledatabase/oracledatabase.go#L90-L93
> 
> What I did was moving the code that was repeated in components and standardizing it for all state stores that are transactional.
> 
> About Resiliency:
> 
> * In the case of BulkSet and BulkDelete, nothing changes actually, because before Resiliency was applied to the "BulkSet" invocation and not each individual Set. The loop happened inside the DefaultBulkStore that was unaware of resiliency. This would actually create some interesting issues when not using transactions and doing a BulkSet with Etags, if some items are set and some are not
> * In the case of BulkGet there is a behavior change with Resiliency as explained in the first message.

This is actually a breaking change. When using bulk operations, users are accepting/expecting partial success. Enforcing transactional changes that expectation. Regarding performance improvement, it might not be the case compared to a native bulk operation - we would need to compare - regardless, performance improvement does not matter if the behavior is not what is expected. I vote for not merging this behavior change into the runtime.

**Comment by ItalyPaleAle** [comment]
If the concern is with Bulk Set / Bulk Delete, this is something that needs to change in components-contrib and not in this PR. So this PR should be able to be merged regardless (and it's a pretty urgent PR , as it's blocking pinning components-contrib and [blocking contributors](https://github.com/dapr/components-contrib/issues/2779))

In current master, `BulkSet` and `BulkDelete` are delegating the entire operation to the components:

https://github.com/dapr/dapr/blob/0fec7ad0cb357bade8067a658fb1d46823e1fa0d/pkg/grpc/api.go#L767

**This hasn't changed.**

What has changed is in components-contrib.

In Dapr 1.10, components that did not implement a native "BulkSet" (or "BulkDelete") used the DefaultBulkStore:

https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/store.go#L76-L85

In there, you can see the default action was to perform all the Set operations in sequence (so an actual bulk). Notably, because code in components-contrib is unaware of resiliency, the first time an operation failed, the method would return and the entire call to BulkSet would return an error.

- For example, assume that you are setting 3 values: A, B, C.
- B fails to be written, so BulkSet returns back to the runtime and C isn't attempted.
- The runtime is unaware of what has been written already, so if retries are enabled, it will retry writing A, B, and C.
- Assuming that there's an ETag on A, it will fail on A (because it has been updated but the runtime doesn't have the new etag), and B and C won't be attempted.
- The runtime may try again, and it will fail on A again.

This behavior is not only incorrect, but also unreliable and wasteful.

**Most importantly**, many components that were transactional in Dapr 1.10 already implemented "native" BulkSet and BulkDelete transactionally:


- MySQL: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/mysql/mysql.go#L578-L602
- Postgres: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/postgresql/postgresdbaccess.go#L231-L255
- SQLite: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlite/sqlite.go#L105-L114
- SQL Server: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlserver/sqlserver.go#L615-L634
- Oracle: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/oracledatabase/oracledatabase.go#L90-L93

What I did (in **components-contrib and not here**) was to take the code they were using and sharing and move it to the default bulk store. So now, if the state store is transactional, the default bulk store attempts to use transactions:

https://github.com/dapr/components-contrib/blob/0221ad7edff3f33cb28fa6c7e6d8dcaee44bd5fb/state/bulk.go#L90-L107

## TLDR

Bottom line is, if there are concerns on using BulkSet and BulkDelete with backing transactions, we can make the requested changes, but they are independent from components-contrib and they are not new anyways (at least not for half a dozen components)

**Comment by artursouza** [comment]
> If the concern is with Bulk Set / Bulk Delete, this is something that needs to change in components-contrib and not in this PR. So this PR should be able to be merged regardless (and it's a pretty urgent PR , as it's blocking pinning components-contrib and [blocking contributors](https://github.com/dapr/components-contrib/issues/2779))
> 
> In current master, `BulkSet` and `BulkDelete` are delegating the entire operation to the components:
> 
> https://github.com/dapr/dapr/blob/0fec7ad0cb357bade8067a658fb1d46823e1fa0d/pkg/grpc/api.go#L767
> 
> **This hasn't changed.**
> 
> What has changed is in components-contrib.
> 
> In Dapr 1.10, components that did not implement a native "BulkSet" (or "BulkDelete") used the DefaultBulkStore:
> 
> https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/store.go#L76-L85
> 
> In there, you can see the default action was to perform all the Set operations in sequence (so an actual bulk). Notably, because code in components-contrib is unaware of resiliency, the first time an operation failed, the method would return and the entire call to BulkSet would return an error.
> 
> * For example, assume that you are setting 3 values: A, B, C.
> * B fails to be written, so BulkSet returns back to the runtime and C isn't attempted.
> * The runtime is unaware of what has been written already, so if retries are enabled, it will retry writing A, B, and C.
> * Assuming that there's an ETag on A, it will fail on A (because it has been updated but the runtime doesn't have the new etag), and B and C won't be attempted.
> * The runtime may try again, and it will fail on A again.
> 
> This behavior is not only incorrect, but also unreliable and wasteful.
> 
> **Most importantly**, many components that were transactional in Dapr 1.10 already implemented "native" BulkSet and BulkDelete transactionally:
> 
> * MySQL: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/mysql/mysql.go#L578-L602
> * Postgres: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/postgresql/postgresdbaccess.go#L231-L255
> * SQLite: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlite/sqlite.go#L105-L114
> * SQL Server: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/sqlserver/sqlserver.go#L615-L634
> * Oracle: https://github.com/dapr/components-contrib/blob/a0041cd51767d07bf194d903e7cb9d03e0b15e55/state/oracledatabase/oracledatabase.go#L90-L93
> 
> What I did (in **components-contrib and not here**) was to take the code they were using and sharing and move it to the default bulk store. So now, if the state store is transactional, the default bulk store attempts to use transactions:
> 
> https://github.com/dapr/components-contrib/blob/0221ad7edff3f33cb28fa6c7e6d8dcaee44bd5fb/state/bulk.go#L90-L107
> 
> ## TLDR
> Bottom line is, if there are concerns on using BulkSet and BulkDelete with backing transactions, we can make the requested changes, but:
> 
> * It's not a new thing (at least not for the components that are already doing that in 1.10)
> * It's a change in components-contrib and not runtime

That behavior can and should be changed since the response is itemized, so the runtime should know which of the entries failed and only retry those.

File: dapr/proto/components/v1/state.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (artursouza) [reply dapr/proto/components/v1/state.proto:267]: Does this mean pluggable components will not work anymore? It was not clear if this is just not needed or the PR is just removing it without dealing with the consequences.

**Comment by ItalyPaleAle** [comment]
> That behavior can and should be changed since the response is itemized, so the runtime should know which of the entries failed and only retry those.

We can change that by removing `BulkSet` and `BulkDelete` from the interface in components-contrib, and making the runtime perform calls 1-by-1. This way the runtime knows what items are successful.

We can do that in 2 follow-up PRs:

- First a PR in runtime to stop using the `BulkSet` and `BulkDelete` method in components
- Next a PR in contrib to remove those methods from the interface (if we do this after, at least we won't break pinning again)

File: dapr/proto/components/v1/state.proto (hunk 2 above)
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply dapr/proto/components/v1/state.proto:267]: They will need to implement BulkGet, while before they could avoid implementing the business logic and return `false` here. So all pluggable state stores will need to be updated.

It's my understanding that the interfaces for pluggable components are not stable and can be broken when the interfaces in components-contrib change.

File: pkg/actors/actors_test.go [hunk 3]
```diff
@@ -904,18 +734,6 @@ func TestCreateReminder(t *testing.T) {
 	assert.Equal(t, numReminders, len(reminderReferences))
 	assert.Equal(t, numReminders, len(reminders))
 
-	// For the rest of the test, we're disabling bulk get in the fake state store
-	store, ok := testActorsRuntime.compStore.GetStateStore("actorStore")
```
Comment (artursouza) [inline pkg/actors/actors_test.go:908]: Why is this removed?

**Review by artursouza — CHANGES_REQUESTED** [review]

File: pkg/actors/actors_test.go (hunk 3 above)
Comment (ItalyPaleAle) [reply pkg/actors/actors_test.go:908]: All state stores now implement `BulkGet`. This test was asserting for state stores that returned `false` from `BulkGet`, so it's unnecessary.

File: pkg/components/state/pluggable.go [hunk 4]
```diff
@@ -265,7 +268,7 @@ func (ss *grpcStateStore) BulkGet(ctx context.Context, req []state.GetRequest) (
 			ContentType: strNilIfEmpty(resp.ContentType),
 		}
 	}
-	return bulkGetResponse.Got, items, nil
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/components/state/pluggable.go:268]: (Aside from the fact that it was an odd API design choice; should have been an optional interface method instead). All state store components now implement `BulkGet`, either natively or through the default bulk store in components-contrib. So this isn't needed.

The "looping" across all keys has been moved to components-contrib, since before it was implemented 3 times in the runtime, not consistently (causing bugs like those fixed in #6134). One of the 3 instances had a TODO (with your name :) ) to move the code to components-contrib.

File: pkg/grpc/api_test.go [hunk 5]
```diff
@@ -3062,39 +3062,7 @@ func TestStateAPIWithResiliency(t *testing.T) {
 			Keys:      []string{"failingBulkGetKey", "goodBulkGetKey"},
 		})
 
-		assert.NoError(t, err)
-		assert.Equal(t, 2, failingStore.Failure.CallCount("failingBulkGetKey"))
-		assert.Equal(t, 1, failingStore.Failure.CallCount("goodBulkGetKey"))
-	})
-
-	t.Run("bulk state get times out on single with resiliency", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/grpc/api_test.go:3070]: Because BulkGet is now implemented in components-contrib, it's not aware of item-specific resiliency. It will retry all GET operations at once. This is necessary to make sure that components can (although don't have to) implement a "native" BulkGet that performs a single round-trip to the database (for example, SQL databases now can perform a single "SELECT" query with multiple keys, which was not possible before).

File: pkg/http/api_test.go [hunk 6]
```diff
@@ -4143,39 +4145,7 @@ func TestV1StateEndpoints(t *testing.T) {
 
 		resp := fakeServer.DoRequest("POST", apiPath, body, nil)
 
-		assert.Equal(t, 200, resp.StatusCode)
-		assert.Equal(t, 2, failingStore.Failure.CallCount("failingBulkGetKey"))
-		assert.Equal(t, 1, failingStore.Failure.CallCount("goodBulkGetKey"))
-	})
-
-	t.Run("bulk state get times out on single with resiliency", func(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/http/api_test.go:4151]: Same as above

File: pkg/proto/components/v1/state.pb.go [hunk 7]
```diff
@@ -1419,18 +1419,66 @@ func (*BulkDeleteResponse) Descriptor() ([]byte, []int) {
 	return file_dapr_proto_components_v1_state_proto_rawDescGZIP(), []int{20}
 }
 
+type BulkGetRequestOptions struct {
+	state         protoimpl.MessageState
+	sizeCache     protoimpl.SizeCache
+	unknownFields protoimpl.UnknownFields
+
+	Parallelism int64 `protobuf:"varint,1,opt,name=parallelism,proto3" json:"parallelism,omitempty"`
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/proto/components/v1/state.pb.go:1427]: Yes: https://github.com/dapr/components-contrib/blob/master/state/bulk.go#L22-L27

Although components that implement native BulkGet can decide to ignore that (because they will perform a single round-trip call to the DB).

File: pkg/grpc/api.go [hunk 8]
```diff
@@ -718,20 +674,27 @@ func (a *api) SaveState(ctx context.Context, in *runtimev1pb.SaveStateRequest) (
...
 		}
 		req := state.SetRequest{
 			Key:      key,
 			Metadata: s.Metadata,
 		}
 
-		if contentType, ok := req.Metadata[contribMetadata.ContentType]; ok && contentType == contenttype.JSONContentType {
-			if err1 = json.Unmarshal(s.Value, &req.Value); err1 != nil {
-				return empty, err1
+		if req.Metadata[contribMetadata.ContentType] == contenttype.JSONContentType {
```
[the remark this answers is not in the document]
Comment (ItalyPaleAle) [reply pkg/grpc/api.go:694]: When getting from a map such as:

```go
v, ok := mymap[key]
```

If the key doesn't exist, `ok` is `false` and `v` is the zero value. The reason for the `ok` is to detect when the key is present BUT its value is the zero value.

In this case `ok && contentType == contenttype.JSONContentType` was a tautology. Whether the key was missing, or present but with a zero value, it would have failed `== contenttype.JSONContentType` regardless.

**Comment by artursouza** [comment]
> > That behavior can and should be changed since the response is itemized, so the runtime should know which of the entries failed and only retry those.
> 
> We can change that by removing `BulkSet` and `BulkDelete` from the interface in components-contrib, and making the runtime perform calls 1-by-1. This way the runtime knows what items are successful.
> 
> We can do that in 2 follow-up PRs:
> 
> * First a PR in runtime to stop using the `BulkSet` and `BulkDelete` method in components
> * Next a PR in contrib to remove those methods from the interface (if we do this after, at least we won't break pinning again)


We should keep the BulkSet and BulkDelete but change the response to be itemized, like in BulkGet.

**Review by artursouza — APPROVED** [review]
Approved to unblock merged from contrib to runtime but we need to change the default behavior of bulk to not be transactional and allow itemized responses.

**Comment by berndverst** [comment]
This caused a regression which changes the gRPC error codes for things like ETAG mismatch. It breaks the Python E2E test "state_store" (in examples folder)

**Comment by berndverst** [comment]
To be clear on ETAG Mismatch over `gRPC` the error code `ABORTED` was expected, but now this switched to `INTERNAL`.

See here the error that is expected to be passed through: https://github.com/dapr/dapr/blob/master/pkg/grpc/api.go#L756

**Comment by berndverst** [comment]
Relate to this PR: https://github.com/dapr/components-contrib/pull/2747
